python/graphstorm/model/gnn_encoder_base.py

```diff
@@ -210,25 +232,54 @@ def dist_inference_one_layer(layer_id, g, dataloader, target_ntypes, layer, get_
     -------
         dict of Tensors : the inferenced tensors.
     """
+    len_dataloader = max_num_batch = len(list(dataloader))
+    tensor = th.tensor([len_dataloader], device=device)
+    if is_distributed():
+        th.distributed.all_reduce(tensor, op=th.distributed.ReduceOp.MAX)
+        max_num_batch = tensor[0]
+
+    dataloader_iter = iter(dataloader)
     y = {}
-    for iter_l, (input_nodes, output_nodes, blocks) in enumerate(dataloader):
+
+    # WholeGraph does not support imbalanced batch numbers across processes/trainers
+    # TODO (IN): Fix dataloader to have same number of minibatches.
+    for iter_l in range(max_num_batch):
+        tmp_keys = []
+        if iter_l < len_dataloader:
+            input_nodes, output_nodes, blocks = next(dataloader_iter)
+            if not isinstance(input_nodes, dict):
+                # This happens on a homogeneous graph.
+                assert len(g.ntypes) == 1
+                input_nodes = {g.ntypes[0]: input_nodes}
+            if not isinstance(output_nodes, dict):
+                # This happens on a homogeneous graph.
+                assert len(g.ntypes) == 1
+                output_nodes = {g.ntypes[0]: output_nodes}
+            tmp_keys = [ntype for ntype in g.ntypes if ntype not in input_nodes]
```

Comment:
if layer_id > 0 we also do not need to do L258-L260

Reply:
Right! Fixed.